### ISSUE TITLE
Avoid `Dispatch::current()`

### ISSUE THREAD

**Issue by carllerche** [hawkw/tokio-trace-prototype#84]
This requires an arc bump in the hot path.

**Comment by hawkw** [comment]
@carllerche I took a crack at replacing `Dispatch::current` with a function that runs a closure on the current dispatcher using `CURRENT_DISPATCH.with(...)`, and only clones the dispatcher when it's necessary to do so to construct a new enabled span. I've [pushed a branch](https://github.com/hawkw/tokio-trace-prototype/compare/eliza/less-clones?expand=1) for these changes.

However, it seems like this may actually have a slight _negative_ impact on span creation, at least on my machine (although both are within the margin of error):

```
running 4 tests
test bench_1_atomic_load              ... bench:           0 ns/iter (+/- 0)
test bench_costly_field_no_subscriber ... bench:          39 ns/iter (+/- 10)
test bench_no_span_no_subscriber      ... bench:           0 ns/iter (+/- 0)
test bench_span_no_subscriber         ... bench:          37 ns/iter (+/- 10)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out

     Running target/release/deps/subscriber-20286345c98bd6a5

running 4 tests
test span_no_fields            ... bench:          92 ns/iter (+/- 29)
test span_repeatedly           ... bench:      18,801 ns/iter (+/- 4,911)
test span_with_fields          ... bench:         189 ns/iter (+/- 51)
test span_with_fields_add_data ... bench:         746 ns/iter (+/- 204)
```

vs. master:

```
     Running target/release/deps/no_subscriber-b05dc9bbe95b9d88

running 4 tests
test bench_1_atomic_load              ... bench:           0 ns/iter (+/- 0)
test bench_costly_field_no_subscriber ... bench:          38 ns/iter (+/- 12)
test bench_no_span_no_subscriber      ... bench:           0 ns/iter (+/- 0)
test bench_span_no_subscriber         ... bench:          33 ns/iter (+/- 8)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out

     Running target/release/deps/subscriber-20286345c98bd6a5

running 4 tests
test span_no_fields            ... bench:          69 ns/iter (+/- 8)
test span_repeatedly           ... bench:      15,417 ns/iter (+/- 1,998)
test span_with_fields          ... bench:         154 ns/iter (+/- 26)
test span_with_fields_add_data ... bench:         689 ns/iter (+/- 121)
```

I'm not sure, but I wonder if this is because executing code inside a `LocalKey::with` has a bigger performance impact than an arc bump? I've noticed in the past that limiting the scope of `LocalKey::with`s as tightly as possible improves performance slightly.

Any insights here?

**Comment by hawkw** [comment]
It seems like the most important part of this issue was that we shouldn't include `Dispatch::current()` _in the public API_, which was addressed by #85. Removing the arc bump is now an implementation detail rather than a breaking change.
